File: frontend/projects/irrtirer/src/app/shared/mosaic-generator/color-compatibility/polygon-rasterization-tool.ts
//Based on https://alienryderflex.com/polygon_fill/
import { Rect } from '../../../core/models/rect.model';
import { Size } from '../../../core/models/size.interface';
import { Vector } from '../../../core/models/vector.model';

export class PolygonRasterizationTool {
    private imageLeftBorder: number = 0;

    private readonly imageSize: Size;
    private readonly imagePosition: Rect;

    public constructor(imagePosition: Rect, imageSize: Size) {
        this.imagePosition = imagePosition;
        this.imageSize = imageSize;
    }

    public rasterizePolygonInRange(vertices: Vector[]): boolean[] {
        const verticesInPixelSpace: Vector[] = this.translateVerticesToPixelSpace(vertices);
        return this.prepareMask(verticesInPixelSpace, this.imageSize);
    }

    private translateVerticesToPixelSpace(vertices: Vector[]): Vector[] {
        const translated: Vector[] = new Array(vertices.length);
        for (let i = 0; i < vertices.length; i++) {
            translated[i] = new Vector(
                ((vertices[i].x - this.imagePosition.x) / this.imagePosition.width) * this.imageSize.width,
                ((vertices[i].y - this.imagePosition.y) / this.imagePosition.height) * this.imageSize.height
            );
        }

        return translated;
    }

    private prepareMask(vertices: Vector[], imageSize: Size): boolean[] {
        const mask: boolean[] = new Array(imageSize.width * imageSize.height);
        const nodeX: number[] = new Array(vertices.length);

        //  Loop through the rows of the image.
        for (let pixelY = 0; pixelY < imageSize.height; ++pixelY) {
            const nodesInRowCount: number = this.buildArrayOfNodes(vertices, nodeX, pixelY);
            this.sortNodes(nodeX, nodesInRowCount);

            //  Fill the pixels between node pairs.
            for (let i = 0; i < nodesInRowCount; i += 2) {
                if (nodeX[i] >= imageSize.width) break;

                if (nodeX[i + 1] > this.imageLeftBorder) {
                    if (nodeX[i] < this.imageLeftBorder) {
                        nodeX[i] = this.imageLeftBorder;
                    }
                    if (nodeX[i + 1] > imageSize.width) {
                        nodeX[i + 1] = imageSize.width;
                    }

                    for (let pixelX = nodeX[i]; pixelX < nodeX[i + 1]; pixelX++) {
                        mask[pixelX + pixelY * imageSize.width] = true;
                    }
                }
            }
        }

        return mask;
    }

    private buildArrayOfNodes(vertices: Vector[], nodeX: number[], pixelY: number): number {
        let nodesInRowCount = 0;
        for (let i = 0, j = vertices.length - 1; i < vertices.length; i++) {
            if ((vertices[i].y < pixelY && vertices[j].y >= pixelY) || (vertices[j].y < pixelY && vertices[i].y >= pixelY)) {
                nodeX[nodesInRowCount++] =
                    vertices[i].x + ((pixelY - vertices[i].y) / (vertices[j].y - vertices[i].y)) * (vertices[j].x - vertices[i].x);
            }
            j = i;
        }

        return nodesInRowCount;
    }

    private sortNodes(nodeX: number[], nodesInRowCount: number): void {
        let i = 0;
        while (i < nodesInRowCount - 1) {
            if (nodeX[i] > nodeX[i + 1]) {
                const swapValue: number = nodeX[i];
                nodeX[i] = nodeX[i + 1];
                nodeX[i + 1] = swapValue;
                if (i != 0) {
                    i--;
                }
            } else {
                i++;
            }
        }
    }
}

File: backend/Irrtirer/Irrtirer.Generator/ColorCompatibility/PolygonRasterizationTool.cs
//Based on https://alienryderflex.com/polygon_fill/

using System.Drawing;
using System.Numerics;

namespace Irrtirer.Generator.ColorCompatibility
{
    internal class PolygonRasterizationTool
    {
        private const int ImageLeftBorder = 0;

        private readonly Size imageSize;
        private readonly RectangleF imagePosition;

        public PolygonRasterizationTool(RectangleF imagePosition, Size imageSize)
        {
            this.imagePosition = imagePosition;
            this.imageSize = imageSize;
        }

        public bool[] RasterizePolygonInRange(Vector2[] vertices)
        {
            Vector2[] verticesInPixelSpace = TranslateVerticesToPixelSpace(vertices);
            return PrepareMask(verticesInPixelSpace, imageSize);
        }

        private Vector2[] TranslateVerticesToPixelSpace(Vector2[] vertices)
        {
            Vector2[] translated = new Vector2[vertices.Length];
            for (int i = 0; i < vertices.Length; i++)
            {
                translated[i] = new Vector2((vertices[i].X - imagePosition.X) / imagePosition.Width * imageSize.Width,
                                            (vertices[i].Y - imagePosition.Y) / imagePosition.Height * imageSize.Height);
            }

            return translated;
        }

        private bool[] PrepareMask(Vector2[] vertices, Size imageSize)
        {
            bool[] mask = new bool[imageSize.Width * imageSize.Height];

            int[] nodeX = new int[vertices.Length];

            //  Loop through the rows of the image.
            for (int pixelY = 0; pixelY < imageSize.Height; ++pixelY)
            {
                BuildArrayOfNodes(vertices, nodeX, out int nodesInRowCount, pixelY);
                SortNodes(nodeX, nodesInRowCount);

                //  Fill the pixels between node pairs.
                for (int i = 0; i < nodesInRowCount; i += 2)
                {
                    if (nodeX[i] >= imageSize.Width)
                        break;

                    if (nodeX[i + 1] > ImageLeftBorder)
                    {
                        if (nodeX[i] < ImageLeftBorder)
                            nodeX[i] = ImageLeftBorder;
                        if (nodeX[i + 1] > imageSize.Width)
                            nodeX[i + 1] = imageSize.Width;

                        for (int pixelX = nodeX[i]; pixelX < nodeX[i + 1]; pixelX++)
                        {
                            mask[pixelX + pixelY * imageSize.Width] = true;
                        }
                    }
                }
            }

            return mask;
        }

        private static void BuildArrayOfNodes(Vector2[] vertices, int[] nodeX, out int nodesInRowCount, int pixelY)
        {
            nodesInRowCount = 0;
            for (int i = 0, j = vertices.Length - 1; i < vertices.Length; i++)
            {
                if (vertices[i].Y < pixelY && vertices[j].Y >= pixelY || vertices[j].Y < pixelY && vertices[i].Y >= pixelY)
                {
                    nodeX[nodesInRowCount++] = (int)(vertices[i].X + (pixelY - vertices[i].Y) / (vertices[j].Y - vertices[i].Y) * (vertices[j].X - vertices[i].X));
                }
                j = i;
            }
        }

        private static void SortNodes(int[] nodeX, int nodesInRowCount)
        {
            int i = 0;
            while (i < nodesInRowCount - 1)
            {
                if (nodeX[i] > nodeX[i + 1])
                {
                    int swapValue = nodeX[i];
                    nodeX[i] = nodeX[i + 1];
                    nodeX[i + 1] = swapValue;
                    if (i != 0)
                    {
                        i--;
                    }
                }
                else
                {
                    i++;
                }
            }
        }
    }
}
